aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip

```diff
@@ -206,102 +306,51 @@ mha_fwd(const at::Tensor &q,         // batch_size x seqlen_q x num_heads x head
-                    v_t.stride(0), v_t.stride(1), v_t.stride(2), v_t.stride(3),                                             \
-                    output_t.stride(0), output_t.stride(1), output_t.stride(2), output_t.stride(3),                         \
-                    q_t.sizes()[0], q_t.sizes()[1], seqlen_q, seqlen_k, p_dropout,                                          \
-                    *(uint64_t*)(seed_t.data_ptr()), *(uint32_t*)(offset_t.data_ptr()),                                     \
-                    (FP*)(softmax_fa_t.data_ptr()),                                                                         \
-                    stream);                                                                                                \
-  } while(0)
-
-  // TODO: Ugly but works
-  constexpr int kFwdUseCausal = 3;
-  constexpr int kFwdNoCausal  = 1;
-  int d_head = q_t.sizes()[3];
-  constexpr int BM = BLOCK_M;
-  constexpr int BN = BLOCK_N;
-  if (q_dtype == at::kHalf) {
-    if (is_causal) {
-      if (d_head == 16)
-        CALL_FWD(__fp16,kFwdUseCausal,BM,16,BN,true,true,true);
-      else if (d_head == 32)
-        CALL_FWD(__fp16,kFwdUseCausal,BM,32,BN,true,true,true);
-      else if (d_head == 64)
-        CALL_FWD(__fp16,kFwdUseCausal,BM,64,BN,true,true,true);
-      else if (d_head == 128)
-        CALL_FWD(__fp16,kFwdUseCausal,BM,128,BN,true,true,true);
-    } else {
-      if (d_head == 16)
-        CALL_FWD(__fp16,kFwdNoCausal,BM,16,BN,true,true,true);
-      else if (d_head == 32)
-        CALL_FWD(__fp16,kFwdNoCausal,BM,32,BN,true,true,true);
-      else if (d_head == 64)
-        CALL_FWD(__fp16,kFwdNoCausal,BM,64,BN,true,true,true);
-      else if (d_head == 128)
-        CALL_FWD(__fp16,kFwdNoCausal,BM,128,BN,true,true,true);
-    }
-  } else if (q_dtype == at::kBFloat16) {
-    if (is_causal) {
-      if (d_head == 16)
-        CALL_FWD(__bf16,kFwdUseCausal,BM,16,BN,true,true,true);
-      else if (d_head == 32)
-        CALL_FWD(__bf16,kFwdUseCausal,BM,32,BN,true,true,true);
-      else if (d_head == 64)
-        CALL_FWD(__bf16,kFwdUseCausal,BM,64,BN,true,true,true);
-      else if (d_head == 128)
-        CALL_FWD(__bf16,kFwdUseCausal,BM,128,BN,true,true,true);
-    } else {
-      if (d_head == 16)
-        CALL_FWD(__bf16,kFwdNoCausal,BM,16,BN,true,true,true);
-      else if (d_head == 32)
-        CALL_FWD(__bf16,kFwdNoCausal,BM,32,BN,true,true,true);
-      else if (d_head == 64)
-        CALL_FWD(__bf16,kFwdNoCausal,BM,64,BN,true,true,true);
-      else if (d_head == 128)
-        CALL_FWD(__bf16,kFwdNoCausal,BM,128,BN,true,true,true);
-    }
+  at::Tensor softmax_fa_t;
+  if (return_softmax) {
+    softmax_fa_t = at::empty({batch_size, num_heads, seqlen_q, seqlen_k},
+                             at::dtype(q.dtype()).device(q.device()));
+  } else {
+    softmax_fa_t = at::empty({ 0, 0, 0, 0 }, at::dtype(q.dtype()).device(q.device()));
   }
 
-  //undo reorder tensors
```

Comment:
do you no longer need to permute the output?

Reply:
Yes we don't, the only limit is the last dimension must be continuous, which is also enforced by CUTLASS implementation.